### ISSUE TITLE
Add support for local variables

### ISSUE THREAD

**Issue by tronical** [slint-ui/slint#2752]
### Discussed in https://github.com/slint-ui/slint/discussions/2751

<div type='discussions-op-text'>

<sup>Originally posted by **loloof64** May 18, 2023</sup>
Hi, is it possible to define local variables in ui scripts components/functions ?

I mean something like this for example

```javascript
export component MySquare inherits Rectangle {
    width: 100px;
    height: 100px;
    background: red;
    in property<string> items: "abcd/efgh/ijkl/mnop";

    public function get_items_arr() -> [string] {
        let arr = [];
        for elt in items {
            if (elt != "/") arr.push(elt);
        }
        arr
    }
}
```</div>

**Comment by ogoffart** [comment]
Local variables might make some sense. But i would be hesitant to implement `for` loop.
Our strategy had been to keep the language intentionally simple because the actual logic should be written in native code.

**Comment by ogoffart** [comment]
We would need to decide a syntax for it.

I'd go with `let` as this is the same syntax as rust and typescript.

```
let foo : string = "123";
```

(type annotation optional)

**Comment by Tmpod** [comment]
I agree with this. Having variables is useful sometimes, but Slint shouldn't really go deeper than that, and let the backend language handle that.

**Comment by FloVanGH** [comment]
The code for the timer picker #5087 would also profit from local variables.

**Comment by tsaxking** [comment]
> We would need to decide a syntax for it.
> 
> I'd go with `let` as this is the same syntax as rust and typescript.
> 
> ```
> let foo : string = "123";
> ```
> 
> (type annotation optional)

Has there been any update on this feature? I would love this!

**Comment by Gibbz** [comment]
I was struggling with this today too! As a new user I couldn't figure this out, and didn't realise it was not implemented. 

I'm trying to create a new point, and assign it to an array. I guess this needs to be done in rust instead of slint, then passed back to rust. Which is fine, I just didn't know 👍

**Comment by ogoffart** [comment]
Complex logic with `for` loop should be implemented in callback in native code.

Local variables would be sometimes quite useful to store intermediate result.

A workaround (which is not a replacement for it but can be useful sometimes) is to use an intermediate function:

```slint
function foobar() -> int {
    return foobar-impl(compute-value());
}
function foobar-impl(value: int) -> {
    return use(value, value*2);
}

// Instead of 
// let value = compute-value();
// use(value, value*2)
```

**Comment by ogoffart** [comment]
I'll tag this as good-first-issue to mean that if someone really wants this, they can give it a shot at implementing. It is not trivial, but being localized in the compiler frontend it should be doable by anyone who can write rust code. 

Here are how to do it:
 
 1. Actually parse the `let identifier =`  statement in https://github.com/slint-ui/slint/blob/edbcab7d96943d4f47e53d9078a3d4f45f266a6f/internal/compiler/parser/statements.rs#L18
 2. This should produce a new node kind in the CodeBlock defined around https://github.com/slint-ui/slint/blob/edbcab7d96943d4f47e53d9078a3d4f45f266a6f/internal/compiler/parser.rs#L370
 3. In the resolving pass, that new node  should produce an `Expression::StoreLocalVariable`. probably around there https://github.com/slint-ui/slint/blob/edbcab7d96943d4f47e53d9078a3d4f45f266a6f/internal/compiler/passes/resolving.rs#L188
    and store the type of that expression in the `LookupCtx` for further access.
 4. In lookup.rs, add the code to find that local variable later. https://github.com/slint-ui/slint/blob/edbcab7d96943d4f47e53d9078a3d4f45f266a6f/internal/compiler/lookup.rs#L879
    Should be similar to the existing `ArgumentsLookup`, but generate a `Expression::ReadLocalVariable`
 5. Add tests in tests/cases/expr or tests/cases/lookup or somewhere where it fits. (this step could also be done first, for TDD. What I'd do is just to use it in a small .slint file, and try to run the viewer over it until it works)
 6. Don't forget documentation

That should be it.
